fairseq/modules/moe_layer.py

```diff
@@ -0,0 +1,464 @@
+            shuffle_sort = torch.randperm(features.size(0), device=features.device)
+            features = All2All.apply(features[shuffle_sort])
+
+        with torch.no_grad():
+            # Compute similarity of each token to each expert, for routing
+            token_expert_affinities = features.matmul(self.expert_centroids.transpose(0, 1))
+            # TODO @margsli should be identical to below, is the einsum faster?
+            # check that features and expert_centroids' dimensions are ordered correctly
+            # token_expert_affinities = torch.einsum('sbh,eh->sbe', features, self.expert_centroids)
+            # TODO @margsli this is only argmax, should the other sampling strategies be implemented?
+            self.last_p = torch.softmax(token_expert_affinities, -1)
+
+        # Compute which token goes to which expert
+        routing_strategy = 'greedy'
+        sort_by_expert, input_splits, output_splits = self.assignment(token_expert_affinities, strategy=routing_strategy)
+        # Swap these tokens for the right ones for our expert
+        routed_features = All2All.apply(features[sort_by_expert], output_splits, input_splits)
+
+        if routed_features.size(0) > 0:
+            # Mix in the expert network based on how appropriate it is for these tokens
+            alpha = torch.sigmoid(routed_features.mv(self.expert_centroids[self.expert_id])).unsqueeze(1)
+            routed_features = alpha * self.expert_network(routed_features) + (1 - alpha) * routed_features
+        # Return to original worker and ordering
+        result = All2All.apply(routed_features, input_splits, output_splits)[self.inverse_sort(sort_by_expert)]
+
+        if self.shuffle and is_training:
+            # Undo shuffling
+            result = All2All.apply(result)[self.inverse_sort(shuffle_sort)]
+
+        # Return additional Nones for compatibility with TransformerDecoderLayer
+        return result.view(input_features.size()), None, None
+
+    def inverse_sort(self, order):
+        # Creates an index that undoes a sort: xs==xs[order][inverse_sort(order)]
+        return torch.empty_like(order).scatter_(0, order, torch.arange(0, order.size(0), device=order.device))
+
+    # Assigns each token to the top k experts
+    # NOTE: this ONLY works for k=1 right now b/c of self.last_mask assignment
+    def assignment(self, scores, strategy='greedy', k=1):
+        if strategy == 'greedy':
+            # get a flattened list of the workers to send each token to
+            # scores is s,b,e, token_to_workers is s*b, with the worker # for each token
+            idx = torch.topk(scores, dim=1, k=k, largest=True).indices
+            self.last_mask = F.one_hot(idx[:, 0], num_classes=self.num_workers)
+            token_to_workers = idx.view(-1)
+            # token_to_workers is worker # for each token, sorted, sort_ordering gives original index
+            token_to_workers, sort_ordering = torch.sort(token_to_workers)
```

Comment:
Just to understand what is going on here: You have dimensions of `[b, s]` and now you sort the entire array, that gives you the sorted indices (consecutive tokens to send to each expert) and the indices for that. 

Then you do indices // k to distribute them to workers in the case you have multiple experts per GPU?